dmoj/cptbox/filesystem_policies.py

```diff
@@ -0,0 +1,113 @@
+
+
+class AccessMode(Enum):
+    NONE = 0
+    EXACT = 1
+    RECURSIVE = 2
+
+
+class Dir:
+    def __init__(self):
+        self.access_mode = AccessMode.NONE
+        self.subdirectory_map = {}
+
+
+class File:
+    pass
+
+
+class ExactFile:
+    def __init__(self, path: str):
+        self.path = path
+
+
+class ExactDir:
+    access_mode = AccessMode.EXACT
+
+    def __init__(self, path: str):
+        self.path = path
+
+
+class RecursiveDir:
+    access_mode = AccessMode.RECURSIVE
+
+    def __init__(self, path: str):
+        self.path = path
+
+
+FilesystemAccessRule = Union[ExactFile, ExactDir, RecursiveDir]
+
+
+class FilesystemPolicy:
+    def __init__(self, rules: List[FilesystemAccessRule]):
+        self.root = Dir()
+        for rule in rules:
+            self._add_rule(rule)
+
+    def _add_rule(self, rule: FilesystemAccessRule) -> None:
+        self._assert_rule_type(rule)
+        if rule.path == "/":
+            return self._finalize_root_rule(rule)
+
+        path = rule.path
+        assert os.path.abspath(path) == path, "FilesystemAccessRule must specify a normalized, absolute path to rule"
+        *directory_path, final_component = path.split("/")[1:]
+
+        node = self.root
+        for component in directory_path:
+            new_node = node.subdirectory_map.setdefault(component, Dir())
+            assert isinstance(new_node, Dir), "Cannot add rule: refusing to descend into non-directory"
+            node = new_node
+
+        self._finalize_rule(node, final_component, rule)
+
+    def _assert_rule_type(self, rule: FilesystemAccessRule) -> None:
+        if os.path.exists(rule.path):
+            is_dir = os.path.isdir(rule.path)
+            if isinstance(rule, ExactFile):
+                assert not is_dir, f"Can't apply file rule to directory {rule.path}"
+            else:
+                assert is_dir, f"Can't apply directory rule to non-directory {rule.path}"
+
+    def _finalize_root_rule(self, rule: FilesystemAccessRule) -> None:
+        assert not isinstance(rule, ExactFile), "Root is not a file"
+        self._finalize_directory_rule(self.root, rule)
+
+    def _finalize_rule(self, node: Dir, final_component: str, rule: FilesystemAccessRule) -> None:
+        assert final_component != "", "Must not have trailing slashes in rule path"
+        if isinstance(rule, ExactFile):
+            new_node = node.subdirectory_map.setdefault(final_component, File())
+            assert isinstance(new_node, File), "Can't add ExactFile: Dir rule exists"
+        else:
+            new_node = node.subdirectory_map.setdefault(final_component, Dir())
+            assert isinstance(new_node, Dir), "Can't add rule: File rule exists"
+            self._finalize_directory_rule(new_node, rule)
+
+    def _finalize_directory_rule(self, node: Dir, rule: Union[ExactDir, RecursiveDir]) -> None:
+        node.access_mode = AccessMode(max(node.access_mode.value, rule.access_mode.value))  # Allow the more permissive rule
```

Comment:
We can move this into `AccessMode.most_permissive` to not leak the fact that these are ints here.